Core: Improve project root detection logic

dobesv opened the pull request:
- Support `.hg`
- Support `.yarn`
- Support an env var STORYBOOK_PROJECT_ROOT as an override if other detection methods are not working

Issues: #20314 #14042 #20211

## What I did

Update getProjectRoot logic to cover additional use cases.

## How to test

See the related issues for the cases that lead to this.

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

dobesv committed:
Improve project root detection logic

- Support `.hg`
- Support `.yarn`
- Support an env var STORYBOOK_PROJECT_ROOT as an override if other detection methods are not working

kylegach commented:
Hey, @dobesv! Thank you for this contribution (and congratulations on your first!). Could you add some tests for this, please?

dobesv commented:
Any pointers on how and where to add the tests?  I'm not too familiar with the project, and I don't think there are existing tests for this function unfortunately.  It seems a bit tricky to test.

ndelangen commented on code/lib/core-common/src/utils/paths.ts:
```diff
@@ -1,31 +1,35 @@
-    //
+  // Assume the innermost git/svn/hg folder is the project root
+  const foundVcsMetaDir = ['.git', '.svn', '.hg']
+    .map((dir) => findUp.sync(dir, { type: 'directory' }))
```
This is wasteful, especially considering all these are blocking the main thread sync calls.

The old code before only tried .svn if .git failed.
The new code will try all 3, even if the first one succeeds.

dobesv replied:
True

ndelangen commented on code/lib/core-common/src/utils/paths.ts:
```diff
@@ -1,31 +1,35 @@
+  // Walk upwards out of any node_modules or .yarn folders.  We have ot keep going up multiple steps
+  // because node_modules is commonly nested instead other node_modules or inside the .yarn cache folder
+  let projectRoot = process.cwd();
+  for (;;) {
+    const cwd = projectRoot;
+    const foundNodeModulesDir = ['node_modules', '.yarn']
+      .map((dir) => findUp.sync(dir, { cwd, type: 'directory' }))
+      .filter(Boolean)[0];
+    if (foundNodeModulesDir) {
+      projectRoot = path.join(foundNodeModulesDir, '..');
+    } else {
+      break;
+    }
```
Won't a split on a regex be able to essentially do the same thing, without using findUp?

dobesv replied:
Hmm good point.

ndelangen reviewed (changes requested):
I left 2 comments

ndelangen commented:
@dobesv thank you for contributiong! I think this change is good, but I'd like it implemented in a slightly different way, I hope you can make that happen? 🙏

dobesv committed:
Fixes

- Stop checking for VCS folders if we find one
- Use __dirname for the .yarn/node_modules search
- Also consider `.yarn` along with VCS folders to indicate the project root

dobesv commented:
Thanks for the review @ndelangen .  I made changes based on your feedback.  If you have any thoughts about how I could make automated tests for this without a ton of work, let me know.

ndelangen committed:
Merge branch 'next' into get-project-root-fixup

ndelangen commented:
We could add tests that mock `find-up`; that would mean we only have to specify when find-up with return what string?

ndelangen committed:
Merge branch 'next' into pr/dobesv/20791

ndelangen committed:
Discard changes to .vscode/settings.json

ndelangen committed:
remove duplicate code

ndelangen committed:
Merge branch 'get-project-root-fixup' of github.com:dobesv/storybook into pr/dobesv/20791

ndelangen committed:
Merge branch 'release-8-0' into get-project-root-fixup

ndelangen committed:
fix

ndelangen committed:
Merge branch 'get-project-root-fixup' of github.com:dobesv/storybook into pr/dobesv/20791
